### ISSUE TITLE
[Bug] ConsistentHashingStickyKeyConsumerSelector's getConsumerKeyHashRanges doesn't return correct results

### ISSUE THREAD

**Issue by lhotari** [apache/pulsar#23321]
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Read release policy

- [X] I understand that unsupported versions don't get bug fixes. I will attempt to reproduce the issue on a supported version of Pulsar client and Pulsar broker.

### Version

all versions

### Minimal reproduce step

After the last entry in the hash ring, it wraps around to the first entry:
https://github.com/apache/pulsar/blob/7e73967f9daf368562fa0318d4ecb62272d72174/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ConsistentHashingStickyKeyConsumerSelector.java#L123-L125

This is ignored in the getConsumerKeyHashRanges method:
https://github.com/apache/pulsar/blob/7e73967f9daf368562fa0318d4ecb62272d72174/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ConsistentHashingStickyKeyConsumerSelector.java#L138-L145

Another problem is that the rule to pick one consumer when there are hash collisions is also ignored:
https://github.com/apache/pulsar/blob/7e73967f9daf368562fa0318d4ecb62272d72174/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ConsistentHashingStickyKeyConsumerSelector.java#L127

### What did you expect to see?

the results returned by the getConsumerKeyHashRanges should be accurate and correct

### What did you see instead?

the getConsumerKeyHashRanges logic
* ignores the "hash ring" wrap around after the last entry
* ignores the logic for choosing a single entry when there's a hash collision
  * this is a conflicting goal currently since for hash collisions, the target consumer is selected based on the input hash
    * for hash ranges with multiple consumers, it would be necessary to scan the complete range to find out the actual hash ranges for a consumer

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

**Comment by lhotari** [comment]
I'm pretty sure that `consumerList.get(hash % consumerList.size())` is wrong. It seems that this type of code would make more sense for picking the consumer:
```java
            Map.Entry<Integer, List<Consumer>> ceilingEntry = hashRing.ceilingEntry(hash);
            int entryHash;
            if (ceilingEntry != null) {
                consumerList =  ceilingEntry.getValue();
                entryHash = ceilingEntry.getKey();
            } else {
                consumerList = hashRing.firstEntry().getValue();
                entryHash = hashRing.firstEntry().getKey();
            }

            return consumerList.get(entryHash % consumerList.size());
```

**Comment by lhotari** [comment]
it will be possible to fix this after #23327 is merged.

I have this locally
```java
@Override
    public Map<Consumer, List<Range>> getConsumerKeyHashRanges() {
        Map<Consumer, List<Range>> result = new LinkedHashMap<>();
        rwLock.readLock().lock();
        try {
            if (hashRing.isEmpty()) {
                return result;
            }
            int start = 0;
            int lastKey = 0;
            for (Map.Entry<Integer, HashRingEntry> entry: hashRing.entrySet()) {
                Consumer consumer = entry.getValue().getSelectedConsumer();
                result.computeIfAbsent(consumer, key -> new ArrayList<>())
                            .add(Range.of(start, entry.getKey()));
                lastKey = entry.getKey();
                start = lastKey + 1;
            }
            // Handle wrap-around
            HashRingEntry firstHashRingEntry = hashRing.firstEntry().getValue();
            Consumer firstSelectedConsumer = firstHashRingEntry.getSelectedConsumer();
            List<Range> ranges = result.get(firstSelectedConsumer);
            if (lastKey != rangeSize - 1) {
                ranges.add(Range.of(lastKey + 1, rangeSize - 1));
            }
            int firstKey = hashRing.firstEntry().getKey();
            if (firstKey != 0) {
                ranges.add(0, Range.of(0, firstKey));
            }
        } finally {
            rwLock.readLock().unlock();
        }
        return result;
    }
```

**Comment by lhotari** [comment]
This will be included in #23327 since the implementation was needed for tests in that PR.

**Comment by lhotari** [comment]
Issue #16654 is related.
